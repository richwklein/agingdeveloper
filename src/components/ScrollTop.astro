---
/**
 * @name ScrollTop
 * Component for creating a floating action button to scroll to the top of the page.
 *
 * @return {Function} an Astro component factory function.
 *
 * @example
 * <ScrollTop />
 */
import { Icon } from "astro-icon/components";
---

<scroll-top class:list={["fixed bottom-8 right-8"]}>
  <button
    class:list={[
      "rounded-full bg-secondary-main p-2 text-secondary-contrast opacity-0 shadow-xl hover:bg-secondary-dark",
    ]}
  >
    <Icon name="mdi:keyboard-arrow-up" class:list={["h-6 w-6 "]} />
  </button>
</scroll-top>

<script>
  // Define the behaviour for our new type of HTML element.
  class ScrollTop extends HTMLElement {
    constructor() {
      super();

      const button = this.querySelector("button");
      button?.addEventListener("click", () => {
        window.scrollTo({ top: 0, left: 0, behavior: "smooth" });
      });

      window.onscroll = () => {
        if (document.body.scrollTop > 100 || document.documentElement.scrollTop > 100) {
          button?.classList.remove("fade-out");
          button?.classList.add("fade-in");
        } else {
          button?.classList.remove("fade-in");
          button?.classList.add("fade-out");
        }
      };
    }
  }

  // Tell the browser to use our AstroHeart class for <astro-heart> elements.
  customElements.define("scroll-top", ScrollTop);
</script>
