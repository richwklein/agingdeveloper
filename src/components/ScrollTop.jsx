import React from "react";
import {KeyboardArrowUpOutlined} from "@mui/icons-material";
import {Box, Fab, useScrollTrigger, Zoom} from "@mui/material";

export const ScrollTop = () => {
  const hasScroll = useScrollTrigger({threshold: 100});

  const handleClick = (event) => {
    const anchor = (event.target.ownerDocument).querySelector(
        "#scroll-top",
    );

    if (anchor) {
      anchor.scrollIntoView({behavior: "smooth", block: "center"});
    }
  };

  return (
    <Zoom in={hasScroll}>
      <Box
        onClick={handleClick}
        role="presentation"
        sx={(theme) => ({
          position: "fixed",
          bottom: theme.spacing(3.5),
          right: theme.spacing(3.5),
        })}>
        <Fab size="small" color="secondary" aria-label="scroll back to top">
          <KeyboardArrowUpOutlined />
        </Fab>
      </Box>
    </Zoom>
  );
};

export default ScrollTop;
