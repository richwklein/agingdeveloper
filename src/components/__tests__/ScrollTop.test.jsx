import React from "react";
import * as Mui from "@mui/material";
import {cleanup, render} from "@testing-library/react";
import ScrollTop from "../ScrollTop";

jest.mock("@mui/material", () => {
  return {
    __esModule: true,
    ...jest.requireActual("@mui/material"),
  };
});

describe("ScrollTop", () => {
  const useScrollTrigger = jest.spyOn(Mui, "useScrollTrigger");

  afterEach(() => {
    jest.restoreAllMocks();
    cleanup();
  });

  const renderComponent = () => render(
      <ScrollTop />,
  );

  test("that it renders when scrolled.", () => {
    useScrollTrigger.mockImplementation(() => true);
    const {container} = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("that it does not show up without scroll.", () => {
    useScrollTrigger.mockImplementation(() => false);
    const {container} = renderComponent();
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toContainHTML("visibility: hidden");
  });
});
