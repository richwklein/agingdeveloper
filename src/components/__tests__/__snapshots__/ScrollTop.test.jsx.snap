// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollTop that it does not show up without scroll. 1`] = `
<div
  class="MuiBox-root css-mqrlmh"
  role="presentation"
  style="transform: scale(0); visibility: hidden;"
>
  <button
    aria-label="scroll back to top"
    class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary css-9rrh7h-MuiButtonBase-root-MuiFab-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="KeyboardArrowUpOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`ScrollTop that it renders when scrolled. 1`] = `
<div
  class="MuiBox-root css-mqrlmh"
  role="presentation"
  style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
>
  <button
    aria-label="scroll back to top"
    class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary MuiFab-root MuiFab-circular MuiFab-sizeSmall MuiFab-secondary css-9rrh7h-MuiButtonBase-root-MuiFab-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="KeyboardArrowUpOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7.41 15.41 12 10.83l4.59 4.58L18 14l-6-6-6 6z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
